File: src/main/java/com/dbgit/domain/ProcDefinition.java
package com.dbgit.domain;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Objects;

public record ProcDefinition(
        int objectId,
        String schemaName,
        String name,
        String definition,
        String normalizedDefinition
) {
    public ProcDefinition {
        schemaName = Objects.requireNonNull(schemaName, "schemaName");
        name = Objects.requireNonNull(name, "name");
        definition = definition == null ? "" : definition;
        normalizedDefinition = Objects.requireNonNull(normalizedDefinition, "normalizedDefinition");
    }

    public String fullName() {
        return schemaName + "." + name;
    }

    public String digestHex() {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] h = md.digest(normalizedDefinition.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(h.length * 2);
            for (byte b : h) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: Domain/ProcDefinition.cs
using System.Security.Cryptography;
using System.Text;

namespace Dbgit.Domain;

/// <summary>단일 환경에서 조회한 프로시저/함수 정의.</summary>
public sealed record ProcDefinition(
    int ObjectId,
    string SchemaName,
    string Name,
    string Definition,
    string NormalizedDefinition
)
{
    public string FullName => $"{SchemaName}.{Name}";

    public string Digest()
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(NormalizedDefinition));
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}
